src/org/opendatakit/briefcase/ui/reused/Credentials.java

```diff
@@ -0,0 +1,36 @@
+package org.opendatakit.briefcase.ui.reused;
+
+import java.util.Objects;
+import org.apache.http.util.TextUtils;
+
+public class Credentials {
+  private String username;
+  private String password;
+
+  public Credentials(String username, String password) {
+    this.username = username;
+    this.password = password;
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(username, password);
+  }
+
+  public static Credentials from(String username, String password) {
+    if (!(TextUtils.isEmpty(username) || TextUtils.isEmpty(password))) {
+      return new Credentials(username, password);
+    } else {
+      return null;
```

Comment:
We need to throw a `BriefcaseException` here. Returning null doesn't make sense in a `ValueObject`

Reply:
But, the username and password can be blank right?

Reply:
OK, let's back up a little:
- I think we agree that having a `Credentials` without a password or without username doesn't make sense. Therefore, we won't produce any instance of Credentials if we can't provide a username **and** a password.
- We also have decided that, in Briefcase, we use the `Optional<T>` type to model the absence of some value.
- We have defined the factory `Credentials.from(...)` as something that returns an instance of `Credentials`.

Taking all this into consideration, what would that `return null` mean?

If it's "there's nothing to build, so we don't return anything", returning an `Optional.empty()` would be more appropriate, but the signature prevents us from doing so and it would be a little weird that a factory for a Value object would return `Optional<T>` anyway.

If it's "we can't compute that, so we return null to signify an invalid execution of this factory", then we are inventing a new error management mechanism, where `null` means error. We should avoid doing this kind of things because Java already has an error management mechanism: exceptions.

Another insight would be that returning null is too open for interpretation (is it absence of value, is it a special value that means error, is it something else?), modeling return values with types and throwing exceptions is the way to go if we want to be explicit about our business logic.

Reply:
So how do we handle blank credentials? If we call the factory method with the blank values, an exception would be thrown according to the current implementation. Should we change the way it is called?

Reply:
I thought returning null here made sense since the Option.ofNullable() will handle that

Reply:
> I thought returning null here made sense since the Option.ofNullable() will handle that

Yeah, that would make sense if we didn't have `Optional<T>`. This idea is something we need to spread into the code more, but ideally, we should never use `null` for anything.

Reply:
Some examples:

```java

public static Credentials from(String username, String password) {
  if ((TextUtils.isEmpty(username) || TextUtils.isEmpty(password)))
    throw new IllegalArgumentException("You need to provide non empty username and password");
  return new Credentials(username, password);
}

// and then...

Optional<String> maybeUsername = Optional.ofNullable(username);
Optional<String> maybePassword = Optional.of(password).map(Arrays::toString).filter(String::isEmpty);
// A char[] can't ever be null. That's why we filter for non-empty strings at the end
Optional<Credentials> from = maybeUsername.isPresent() && maybePassword.isPresent()
    ? Optional.of(Credentials.from(maybeUsername.get(), maybePassword.get()))
    : Optional.empty();

return new AggregateServerConnectionConfiguration(url, from);
```